### PR TITLE
Fix BLE I/O cancellation handling

### LINKED ISSUE #450: Subsequent reads hang after read fails with GattRequestRejectedException
I have run into a problem when a read request fails with GattRequestRejectedException. Once the request fails, then all other read requests do not complete until the device is disconnected and reconnected.

I isolated the issue to miss handling the pending flag added in #376 when action block returns false. Throwing GattRequestRejectedException leaves the pending flag set to true and so subsequent reads hang waiting for a response that is never going to arrive, since the original request failed.

I suppose the same thing would happen if the action block threw an exception, but I am not how that should be interpreted.

This is the relevant code:

https://github.com/JuulLabs/kable/blob/daf79f53677a984d2ed0b9c1c4cddadabb704094/core/src/androidMain/kotlin/Connection.kt#L53-L69

### PR DESCRIPTION
**Pull request by twyatt**
Closes #450

### CODE REVIEW

**Review by matejbart — CHANGES_REQUESTED** [review]

File: core/src/appleMain/kotlin/Connection.kt [hunk 1]
```diff
@@ -36,9 +39,12 @@ internal class Connection(
         pending = true
         action.invoke()
```
[the remark this answers is not in the document]
Comment (twyatt) [reply core/src/appleMain/kotlin/Connection.kt]: Good catch! I've swapped the lines and went with your suggested `awaitingResponse` naming. Thanks!

File: core/src/androidMain/kotlin/Connection.kt [hunk 2]
```diff
@@ -55,17 +56,21 @@ internal class Connection(
             // Discard response as we've performed another `execute` without the previous finishing. This happens if a
             // previous `execute` was cancelled after invoking GATT action, but before receiving response from callback
             // channel. See https://github.com/JuulLabs/kable/issues/326 for more details.
-            val response = callback.onResponse.receive()
-            pending = false
+            val response = try {
+                callback.onResponse.receive()
```
[the remark this answers is not in the document]
Comment (twyatt) [reply core/src/androidMain/kotlin/Connection.kt]: > Unfortunately, if a CancellationException is thrown here

I've changed this to use `tryReceive` to essentially "clear" the `onResponse` channel (to discard any awaiting responses). `tryReceive` is not `suspend`ing so (my understanding) is that it wouldn't throw a `CancellationException`.

> the `receive()` call to the channel could be performed in an own coroutine scope & job

I like your suggestion of using a `Job` to grab the response, but I'd like to try a bit more with the current approach, as to not make major changes and (hopefully) this is a slightly simpler approach. 🤷 

I'll use a `Job` if I've exhausted ideas around using a `boolean` flag.

[the remark this answers is not in the document]
Comment (matejbart) [reply core/src/androidMain/kotlin/Connection.kt]: Correct me if I'm wrong, but i think that the usage of `tryReceive` still doesn't guarantee a "discard" of a response in the `onResponse` Channel: it's a non-suspending operation, so if `onResponse` is empty at that time (the case when the previous request's response has still not be sent to `onResponse` which can very likely happen), then `tryReceive` will still immediately return with a failed result. As a consequence, the next time we run into `receive()` we still get the wrong response. `tryReceive` would only help us if the response would be already present in the `Channel` but which is something we cannot assume.

I.e. the deadlocks issue would be solved, but unfortunately the bug that has been fixed in #326 would be re-introduced again. 

A simplified version (without error handling etc) of how we could use a `Deferred` in the `execute` func to basically never cancel a `onReceive` call (for an active ongoing connection):

```
private suspend inline fun execute(request: () -> Boolean) = lock.withLock {
    // try to await a still active deferred to not mix up the responses
    // of previous request attempts
    deferredResponse?.let {
        if (it.isActive) {
            // discard the result of the previous request
            it.await()
        }
    }
    // run the request
    if(!request()) {
        // throw an exception if it has not been sent out successfully
        throw IllegalStateException()
    }
    // prepare to receive a value from the onResponse channel
    val deferred = receiveScope.async {
        onResponse.receive()
    }
    // keep the reference and start the job.
    deferredResponse = deferred
    return@withLock deferred.await()
}
```

where `receiveScope` would be simply the current `Peripheral`'s `scope`.

File: core/src/appleMain/kotlin/Connection.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (matejbart) [reply core/src/appleMain/kotlin/Connection.kt]: looks good 👍

File: core/src/androidMain/kotlin/Connection.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (twyatt) [reply core/src/androidMain/kotlin/Connection.kt]: > Correct me if I'm wrong, but i think that the usage of `tryReceive` still doesn't guarantee a "discard" of a response in the `onResponse` Channel: it's a non-suspending operation, so if `onResponse` is empty at that time (the case when the previous request's response has still not be sent to `onResponse` which can very likely happen), then `tryReceive` will still immediately return with a failed result.

Ya, looking at it again, you're right.

> A simplified version (without error handling etc) of how we could use a `Deferred` in the `execute` func to basically never cancel a `onReceive` call (for an active ongoing connection)

Thanks for this! My initial hesitation with this approach was the added complexity and potential performance impacts. So, I benchmarked the `async` operation (on a Pixel 6a, using `androidx.benchmark:benchmark-junit4`) and the results:

```
495   ns           5 allocs    trace    AsyncBenchmark.async
```

Ultimately, 495ns is negligible (especially compared to the BLE I/O it is managing), and 5 object allocations isn't going to hurt.

After discussing internally with the team, we concur that your approach would be better moving forward. I'll be making the changes and pushing them soon.

Thanks so much for all your help!

[the remark this answers is not in the document]
Comment (matejbart) [reply core/src/androidMain/kotlin/Connection.kt]: I understand your concerns about the complexity, thx for the benchmarks. The changes are good.

**Review by matejbart — APPROVED** [review]

**Review by davidtaylor-juul — APPROVED** [review]
🚀

**Review by cedrickcooke — APPROVED** [review]
New impl feels robust, love it

**Pull request merged**